src/pudl/analysis/demand_mapping.py

```diff
@@ -1,294 +1,561 @@
-            electricity demand in demand_df.
-        time_col (str): Label of the column containing timestamps in
-            demand_df.
-        normed_weights (pandas.DataFrame): Dataframe containing normalized
-            values of the attribute being used to allocate demand
-            geographically, for the intersection of the demand area (column)
-            and the target area (row). Column labels are the IDs of the demand
-            areas (e.g. FERC 714 planning areas). Row labels are the IDs of
-            target geometries whose attributes are being used to allocate
-            demand (e.g. the FIPS IDs of census tracts for population based
-            demand allocation). The demand area IDs found in the column labels
-            must be a subset of the demand area IDs found in the timeseries
-            dataframe demand_df.
-
-        target_ids (list): A list of IDs associated with the target
-            geometries whose attributes are being used to allocate demand (e.g.
-            census tract FIPS IDs). These IDs must be a subset of the IDs
-            found in the row index of normed_weights.
+        disagg_layer (GeoDataframe): Completely disaggregated GeoDataFrame
+        by (str or list): single column or list of columns according to which
+        the constants to be allocated are mentioned (e.g. "Demand" (constant)
+        which needs to be allocated is mapped by "id". So, "id" is the `by`
+        column)
+        allocatees (str or list): single column or list of columns according to which
+        the constants to be allocated are mentioned (e.g. "Demand" (constant)
+        which needs to be allocated is mapped by "id". So, "demand" is the
+        `allocatees` column)
+        allocators (str or list): columns by which attribute is weighted and
+        allocated
+        aggregators (str or list): if empty list, the disaggregated data is
+        returned. If aggregators is mentioned, for example REEDs geometries, the
+        data is aggregated at that level.
 
     Returns:
-        pandas.DataFrame: An electricity demand time series analogous to the
-        input time series, but with demand allocated to the geometries
-        identified by the intermediate_ids.
-
+        geopandas.GeoDataFrame: Disaggregated GeoDataFrame with all the various
+        allocated demand columns, or aggregated by `aggregators`
     """
-    ratio_dict = extract_multiple_tracts_demand_ratios(
-        normed_weights, target_ids)
-    demand_trunc = demand_df[demand_df[demand_id_col].isin(ratio_dict.keys())]
-    out_df = (
-        demand_trunc.pivot_table(
-            index=time_col,
-            columns=demand_id_col,
-            values=demand_col
-        )
-        .fillna(0)
-        .dot(pd.Series(ratio_dict))
-    )
-    return out_df
+    # Allowing for single and multiple allocators,
+    # aggregating columns and allocatees
+    if type(allocators) is not list:
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Using type() instead of isinstance() for a typecheck.](https://app.codacy.com/manual/zaneselvans/pudl/pullRequest?prid=5554119)